internal/beater/java_attacher/os_utils.go

```diff
@@ -0,0 +1,142 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build !windows
+// +build !windows
+
+package javaattacher
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"os/user"
+	"path/filepath"
+	"strconv"
+	"syscall"
+)
+
+func (j *JavaAttacher) setRunAsUser(jvm *jvmDetails, cmd *exec.Cmd) error {
+	currentUser, err := user.Current()
+	if err != nil {
+		// If we cannot get the current user, then we assume apm-server
+		// is not the same as the target JVM user.
+		j.logger.Warnf("failed to get the current user: %v", err)
+	} else {
+		j.logger.Debugf("current user: %v", currentUser)
+	}
+
+	if currentUser.Gid != jvm.gid || currentUser.Uid != jvm.uid {
+		uid, gid, err := parseUserIds(jvm.uid, jvm.gid)
+		if err != nil {
+			return err
+		}
+		cmd.SysProcAttr = &syscall.SysProcAttr{}
+		cmd.SysProcAttr.Credential = &syscall.Credential{Uid: uint32(uid), Gid: uint32(gid)}
+	}
+	return nil
+}
+
+// getAttacherJar finds an attacher jar based on the given uid.
+// In POSIX-compliant systems, it would be an attacher jar owned by the given user and with 0600 access mode.
+// If such was not yet created, this function calls createAttacherTempDir to create it and keep a mapping for it
+func (j *JavaAttacher) getAttacherJar(uid string) string {
+	j.tmpAttacherLock.Lock()
+	defer j.tmpAttacherLock.Unlock()
+	attacherJar, uidMapped := j.uidToAttacherJar[uid]
+	if !uidMapped {
+		tmpAttacherJar, err := j.createAttacherTempDir(uid)
+		if err != nil {
+			j.logger.Errorf("failed to create tmp dir for user %v, using the bundled attacher jar", err)
+		} else {
+			attacherJar = tmpAttacherJar
+		}
+		j.uidToAttacherJar[uid] = attacherJar
+	}
+	return attacherJar
```

Comment:
Yes, much better.
The original form was setting `attacherJar = ""` in case of error, but then I realized there is a zero value for strings, so I removed it an left this redundant variable...